# Required libraries :

In [1]:
import pandas as pd
import numpy as np
import glob, os

<mark>A noter j'ai supprimé les données concernant 2009 car n'a pas la même forme que les autres ( juste pour lire tout en même temps)</mark>

#### Les données sur les caracteristiques :

In [2]:
carac_2005=pd.read_csv('dataset/caracteristiques/caracteristiques_2005.csv',encoding='latin-1')
carac_2005.shape

(87026, 16)

In [3]:
carac_2005.head(2)

Num_Acc  an  mois  jour  hrmn  lum  agg  int  atm  col   com  \
0  200500000001   5     1    12  1900    3    2    1    1    3  11.0   
1  200500000002   5     1    21  1600    1    2    1    1    1  51.0   

            adr gps        lat      long  dep  
0         CD41B   M  5051500.0  294400.0  590  
1  rue de Lille   M  5053700.0  280200.0  590

In [4]:
path =r'dataset/caracteristiques' # use your path
allFiles = glob.glob(path + "/*.csv")
frame = pd.DataFrame()
list_ = []
for file_ in allFiles:
    df = pd.read_csv(file_,encoding='latin-1')
    list_.append(df)
caracteristiques = pd.concat(list_)

In [5]:
caracteristiques.shape

(765576, 16)

##### Les données sur les usagers :

In [6]:
usager_2005=pd.read_csv('dataset/usagers/usagers_2005.csv',encoding='latin-1')
usager_2005.shape

(197498, 12)

In [7]:
usager_2005.head(2)

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  200500000001      1     1     4     1       1    11     0     0      0   
1  200500000001      1     1     3     2       3    11     0     0      0   

   an_nais num_veh  
0   1976.0     A01  
1   1968.0     B02

In [8]:
path =r'dataset/usagers' # use your path
allFiles = glob.glob(path + "/*.csv")
frame = pd.DataFrame()
list_ = []
for file_ in allFiles:
    df = pd.read_csv(file_,encoding='latin-1')
    list_.append(df)
usagers = pd.concat(list_)

In [9]:
usagers.shape

(1710043, 12)

In [10]:
usagers.head(2)

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  200500000001    1.0     1     4     1     1.0  11.0   0.0   0.0    0.0   
1  200500000001    1.0     1     3     2     3.0  11.0   0.0   0.0    0.0   

   an_nais num_veh  
0   1976.0     A01  
1   1968.0     B02

#### Les données sur les lieux :

In [11]:
lieu_2005=df1=pd.read_csv('dataset/lieux/lieux_2005.csv',encoding='latin-1')
lieu_2005.shape

(87026, 18)

In [12]:
lieu_2005.head(2)

Num_Acc  catr  voie   v1   v2  circ  nbv   pr    pr1  vosp  prof  \
0  200500000001   3.0  41.0  0.0    B     2    2  1.0  430.0     0     1   
1  200500000002   2.0  41.0  0.0  NaN     0    2  0.0    0.0     1     1   

   plan  lartpc  larrout  surf  infra  situ  env1  
0     1       0       63     1      0     1     0  
1     1       0      100     1      0     5     0

In [13]:
path =r'dataset/lieux' # use your path
allFiles = glob.glob(path + "/*.csv")
frame = pd.DataFrame()
list_ = []
for file_ in allFiles:
    df = pd.read_csv(file_,encoding='latin-1')
    list_.append(df)
lieux = pd.concat(list_)

C:\Users\pc\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2717: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [14]:
lieux.head(2)

Num_Acc  catr voie   v1   v2  circ  nbv   pr    pr1  vosp  prof  plan  \
0  200500000001   3.0   41  0.0    B   2.0  2.0  1.0  430.0   0.0   1.0   1.0   
1  200500000002   2.0   41  0.0  NaN   0.0  2.0  0.0    0.0   1.0   1.0   1.0   

   lartpc  larrout  surf  infra  situ  env1  
0     0.0     63.0   1.0    0.0   1.0   0.0  
1     0.0    100.0   1.0    0.0   5.0   0.0

In [15]:
lieux.shape

(765576, 18)

#### Les données sur les vehicules :

In [16]:
vehicules_2005=pd.read_csv('dataset/vehicules/vehicules_2005.csv',encoding='latin-1')
vehicules_2005.shape

(149764, 9)

In [17]:
vehicules_2005.head(2)

Num_Acc  senc  catv  occutc  obs  obsm  choc  manv num_veh
0  200500000001     0     7       0    0     2     1     1     A01
1  200500000001     0     7       0    0     2     8    10     B02

In [18]:
vehicules_2005.dtypes

Num_Acc     int64
senc        int64
catv        int64
occutc      int64
obs         int64
obsm        int64
choc        int64
manv        int64
num_veh    object
dtype: object

In [19]:
path =r'dataset/vehicules' # use your path
allFiles = glob.glob(path + "/*.csv")
frame = pd.DataFrame()
list_ = []
for file_ in allFiles:
    df = pd.read_csv(file_,encoding='latin-1')
    list_.append(df)
vehicules = pd.concat(list_)

In [20]:
vehicules.shape

(1307042, 9)

In [21]:
vehicules.head(2)

Num_Acc  senc  catv  occutc  obs  obsm  choc  manv num_veh
0  200500000001   0.0     7       0  0.0   2.0   1.0   1.0     A01
1  200500000001   0.0     7       0  0.0   2.0   8.0  10.0     B02

In [22]:
vehicules.dtypes

Num_Acc      int64
senc       float64
catv         int64
occutc       int64
obs        float64
obsm       float64
choc       float64
manv       float64
num_veh     object
dtype: object

##### Save Data prepared to csv files

In [23]:
#vehicules :
vehicules.to_csv('vehicules_2005-2016.csv')
#lieux
lieux.to_csv('lieux_2005-2016.csv')
#usager :
usagers.to_csv('usager_2005-2016.csv')
#caracteristiques :
caracteristiques.to_csv('caracteristiques_2005-2016.csv')

### Nettoyage des données :

---
####  La table caracteristiques 
---

Vu que la base de données contient assez d'erreur de saisie, il s'avére que l'étape de nettoyage est primordiale avant de passer à d'autre étapes.

In [27]:
# rappel  sur la taille des données :
caracteristiques.shape
# on 16 variables ( ce qui correspond bien au document pdf) et 765 576 accidents.

(765576, 16)

##### Variables catégorielles :

###### gps :

In [28]:

print(len(caracteristiques.gps.unique()))
caracteristiques.gps.unique()
#On voit que gps a 8 valeurs uniques, alors que dans le documents pdf il n'y a que 5 ( M,A,G,R et Y)

9


array(['M', nan, '0', 'A', 'G', 'R', 'S', 'T', 'Y'], dtype=object)

nan, 0 et S se sont considérées comme des valeurs manquantes

In [29]:
caracteristiques[caracteristiques.gps=='S'].shape

(3, 16)

In [30]:
caracteristiques[caracteristiques.gps=='0'].shape

(9, 16)

In [31]:
caracteristiques.gps.isnull().sum()

432956

In [32]:
# puisque cette variable est utilisée pour localiser l'accident, on va remplacer '0' et'S' par nan.
caracteristiques.loc[:,['gps']]=caracteristiques.loc[:,['gps']].replace('0',np.nan)
caracteristiques.loc[:,['gps']]=caracteristiques.loc[:,['gps']].replace('S',np.nan)

In [33]:
#verification :
print(len(caracteristiques.gps.unique()))
caracteristiques.gps.unique()

7


array(['M', nan, 'A', 'G', 'R', 'T', 'Y'], dtype=object)

On garde nan pour l'instant. 

###### jour :

In [34]:
print(len(caracteristiques.jour.unique()))
caracteristiques.jour.unique()
# pas de soucis pour cette variable

31


array([12, 21,  4, 10, 28,  3, 18, 25, 29, 23, 11,  1, 30, 19,  9, 31, 15,
       13, 26,  2,  6,  8, 20, 16,  7,  5, 24, 14, 17, 22, 27], dtype=int64)

###### mois :

In [35]:
print(len(caracteristiques.mois.unique()))
caracteristiques.mois.unique()
# pas de soucis pour cette variable


12


array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12], dtype=int64)

##### an :

In [36]:
print(len(caracteristiques.an.unique()))
caracteristiques.an.unique()
#pareil

11


array([ 5,  6,  7,  8, 10, 11, 12, 13, 14, 15, 16], dtype=int64)

###### lum :

In [37]:
print(len(caracteristiques.lum.unique()))
caracteristiques.lum.unique()
# bon

5


array([3, 1, 5, 4, 2], dtype=int64)

###### agg


In [38]:
print(len(caracteristiques['agg'].unique()))
caracteristiques['agg'].unique()
#bon

2


array([2, 1], dtype=int64)

###### intersection :

In [39]:
print(len(caracteristiques.int.unique()))
caracteristiques.int.unique()

10


array([1, 2, 9, 0, 6, 8, 3, 4, 7, 5], dtype=int64)

In [40]:
# 0 est un intrus 
caracteristiques[caracteristiques.int==0].shape
# il y 106 valeurs de 0
# on va remplacer cette valeur par la valeur la plus courante

(106, 16)

In [41]:
caracteristiques.int.astype(str).describe()
# le plus courant c'est 1 avec une fréquence de 547440
# c'est normale car la majorité des accidents n'ont pas lieu sur les intersections ( vitesse très faible générelement)

count     765576
unique        10
top            1
freq      547440
Name: int, dtype: object

In [42]:
# remplacer 0 par 1: 
caracteristiques.loc[:,['int']]=caracteristiques.loc[:,['int']].replace(0,1)

In [43]:
#verifications :
print(len(caracteristiques.int.unique()))
caracteristiques.int.unique()

9


array([1, 2, 9, 6, 8, 3, 4, 7, 5], dtype=int64)

###### atm :

In [44]:
print(len(caracteristiques.atm.unique()))
caracteristiques.atm.unique()
# il y des nan values, de même on va les remplacer par la valeur la plus courante

10


array([  1.,   2.,   3.,   8.,   9.,   7.,   6.,   4.,   5.,  nan])

In [45]:
# taille des nan values dans atm :
print('taille :', caracteristiques.atm.isnull().sum())
# valeur la plus courante dans atm :
caracteristiques.atm.astype(str).describe()
# 1 est la valeur la plus courante ( normale)

taille : 48


count     765576
unique        10
top          1.0
freq      619775
Name: atm, dtype: object

In [46]:
#remplacer nan par 1 :
caracteristiques.loc[:,['atm']]=caracteristiques.loc[:,['atm']].replace(np.nan,1)

In [47]:
#verification :
print(len(caracteristiques.atm.unique()))
caracteristiques.atm.unique()

9


array([ 1.,  2.,  3.,  8.,  9.,  7.,  6.,  4.,  5.])

###### collision :

In [48]:
print(len(caracteristiques.col.unique()))
caracteristiques.col.unique()
# nan values are in 

8


array([  3.,   1.,   5.,   6.,   2.,   4.,   7.,  nan])

In [49]:
# taille des nan values dans col :
print('taille :', caracteristiques.col.isnull().sum())
# valeur la plus courante dans col :
caracteristiques.col.astype(str).describe()
# 6 est la valeur la plus courante ( autre collision)

taille : 11


count     765576
unique         8
top          6.0
freq      252916
Name: col, dtype: object

In [50]:
#remplacer nan par 6 :
caracteristiques.loc[:,['col']]=caracteristiques.loc[:,['col']].replace(np.nan,6)

In [51]:
#verification :
print(len(caracteristiques.col.unique()))
caracteristiques.col.unique()

7


array([ 3.,  1.,  5.,  6.,  2.,  4.,  7.])

###### Adresse postale :

In [52]:
print(len(caracteristiques.adr.unique()))
caracteristiques.adr.unique()
# des adresse, donc pas de problémes ( sauf erreur de saisie. on va lire adresse par adresse bien sûr pour s'assurer )

337657


array(['CD41B', 'rue de Lille', nan, ..., 'tsoundzou-KOUALE ( rue d',
       "passamanty-M'hogoni (rue", 'kaweni-GEOLE (la)'], dtype=object)

##### Variables numériques :

In [53]:
# valeur nan par variable
caracteristiques.isnull().sum()

Num_Acc         0
an              0
mois            0
jour            0
hrmn            0
lum             0
agg             0
int             0
atm             0
col             0
com             2
adr        127664
gps        432968
lat        436603
long       436607
dep             0
dtype: int64

###### lon/lat

 lat et lon sont des variables numériques. Ils présentent plus de 50% de valeurs manquantes. On les garde comme ça pour l'instant.

###### com :

In [54]:
# cette variable présente 2 valeurs manquantes.
# en voyant les valeurs de cette variable, il s'avère qu'il y beaucoup d'erreur car cette variable est censée d'avoir 3 chiffres alors que
# assez de valeurs sont composées de deux chiffres uniquement
# à voir later comment traiter cette variable
# sinon ca sera pas une variable très importante pour l'étude car ce n'est qu'une variable spatiale.

###### dep

In [55]:
#pas de valeur manquante pour cette variables
print(len(caracteristiques.dep.unique()))
caracteristiques.dep.unique()


101


array([590, 620, 800,  20, 600, 760, 270, 350, 220, 290, 560, 140, 500,
       610, 440, 490, 530, 720, 850, 330, 240, 400, 470, 640, 860, 160,
       170, 790, 870, 190, 230, 310,  90, 120, 320, 460, 650, 810, 690,
       820,  10,  70, 260, 380, 420, 730, 740, 630,  30, 150, 430, 570,
       540, 550, 880, 670, 680, 510,  80, 100, 520, 210, 580, 710, 890,
       250, 390, 700, 900, 130,  40,  50,  60, 830, 840, 340, 110, 300,
       480, 660, 201, 202, 780, 910, 950, 770, 450, 180, 280, 360, 370,
       410, 940, 750, 920, 930, 972, 971, 973, 974, 976], dtype=int64)

In [56]:
dep_values=list(caracteristiques.dep.astype(str).values)

In [57]:
set([dep for dep in dep_values if len(dep)==2])
# on a bien toutes les valeurs correspondantes au dépertement de numéro entre 1 et 9.

{'10', '20', '30', '40', '50', '60', '70', '80', '90'}

In [58]:
print(set([dep for dep in dep_values if len(dep)==3 and dep[-1]!='0' and dep!='202' and dep!='201']))
# 201 

{'974', '972', '973', '971', '976'}


In [59]:
# 201 et 202 remplacent 2A et 2B ( voir lien www.francegene.com/rech-fr/dep-fr.php) : donc pour ces deux valeurs pas de soucis
# mais les autres ils ne se terminent pas par 0
#  normalement il y a 95 departement  d'après le lien ci-dessous, donc 950 doit être la plus grande valeur.
# A voir : soit se sont des nouveau dépertement ? sinon on va les remplacer par la plus courante valeurs.
# later .....

In [60]:
#save table to csv file :
caracteristiques.to_csv('caracteristiques_2005-2016.csv',index=False,)

---
#### la table lieux : 
---

##### variables catégorielles :

In [61]:
# shape :
lieux.shape

(765576, 18)

###### catégorie de route :

In [62]:
print(len(lieux.catr.unique()))
lieux.catr.unique()
# il y des nan values

8


array([  3.,   2.,   4.,   6.,   9.,   5.,   1.,  nan])

In [63]:
# 9 spécifie "autre" de type de route
print(lieux[lieux.catr==9].shape)  
#taille valeur nan
print(lieux.catr.isnull().sum())

(14992, 18)
1


In [64]:
# remplacer nan par 9.( 9 toujours n'est pas connu donc implicitement c'est nan)
lieux.loc[:,['catr']]=lieux.loc[:,['catr']].replace(np.nan,9)

In [65]:
#verification
lieux.catr.isnull().sum()

0

###### régime circulation :

In [66]:
print(len(lieux.circ.unique()))
lieux.circ.unique()
# il y a 0 et nan qui ne sont pas des valeurs pour cette variables

6


array([  2.,   0.,   3.,   4.,   1.,  nan])

In [67]:
print("taille 0",lieux[lieux.circ==0].shape)
print("taille nan",lieux.circ.isnull().sum())

taille 0 (40732, 18)
taille nan 737


In [68]:
lieux.circ.astype(str).describe()
# top est 2( les accidents ont lieu beaucoup sur les  routes nationals)

count     765576
unique         6
top          2.0
freq      485032
Name: circ, dtype: object

In [69]:
# On va remplacer ces deux variables par la plus courante valeur 
lieux.loc[:,['circ']]=lieux.loc[:,['circ']].replace(np.nan,2)
lieux.loc[:,['circ']]=lieux.loc[:,['circ']].replace(0,2)

In [70]:
# verification :
lieux.circ.unique()

array([ 2.,  3.,  4.,  1.])

###### voie réservée :

In [71]:
print(len(lieux.vosp.unique()))
lieux.vosp.unique()
# 0 et nan doivent être changés

5


array([  0.,   1.,   3.,   2.,  nan])

In [72]:
print("taille 0 :",lieux[lieux.vosp==0].shape)
print("taille nan :",lieux.vosp.isnull().sum())
print(lieux.shape)
# on remarque que quasiment toute la colonne est en '0'.
# on remplace nan par 0 et on laisse cette variable.

taille 0 : (717337, 18)
taille nan : 1382
(765576, 18)


In [73]:
lieux.loc[:,['vosp']]=lieux.loc[:,['vosp']].replace(0,np.nan)


###### profil de la route : 

In [74]:
print(len(lieux.prof.unique()))
lieux.prof.unique()

6


array([  1.,   0.,   2.,   3.,   4.,  nan])

In [75]:
print("taille 0 :",lieux[lieux.prof==0].shape)
print("taille nan :",lieux.prof.isnull().sum())
print(lieux.shape)

taille 0 : (58716, 18)
taille nan : 982
(765576, 18)


In [76]:
# on remplace 0 et nan par le mode :
lieux.prof.astype(str).describe()
# top est 1 (Plat)

count     765576
unique         6
top          1.0
freq      578523
Name: prof, dtype: object

In [77]:
lieux.loc[:,['prof']]=lieux.loc[:,['prof']].replace(0,1)
lieux.loc[:,['prof']]=lieux.loc[:,['prof']].replace(np.nan,1)

###### Tracé en plan :
    
    

In [78]:
print(len(lieux.plan.unique()))
lieux.plan.unique()

6


array([  1.,   3.,   2.,   0.,   4.,  nan])

In [79]:
print("taille 0 :",lieux[lieux.plan==0].shape)
print("taille nan :",lieux.plan.isnull().sum())
print(lieux.shape)

taille 0 : (52708, 18)
taille nan : 1001
(765576, 18)


In [80]:
# on remplace 0 et nan par le mode :
lieux.plan.astype(str).describe()
# top est 1 ( Partie rectiligne)

count     765576
unique         6
top          1.0
freq      583560
Name: plan, dtype: object

In [81]:
lieux.loc[:,['plan']]=lieux.loc[:,['plan']].replace(0,1)
lieux.loc[:,['plan']]=lieux.loc[:,['plan']].replace(np.nan,1)

###### Etat de la surface 

In [82]:
print(len(lieux.surf.unique()))
lieux.surf.unique()

11


array([  1.,   2.,   0.,   9.,   7.,   8.,   5.,   6.,   3.,   4.,  nan])

In [83]:
print("taille 0 :",lieux[lieux.surf==0].shape)
print("taille nan :",lieux.surf.isnull().sum())
print(lieux.shape)

taille 0 : (24885, 18)
taille nan : 949
(765576, 18)


In [84]:
# on remplace par 9 (autre: 
lieux.loc[:,['surf']]=lieux.loc[:,['surf']].replace(0,9)
lieux.loc[:,['surf']]=lieux.loc[:,['surf']].replace(np.nan,9)

###### situation de l'acccident

In [85]:
print(len(lieux.situ.unique()))
lieux.situ.unique()

7


array([  1.,   5.,   3.,   4.,   0.,   2.,  nan])

In [86]:
print("taille 0 :",lieux[lieux.situ==0].shape)
print("taille nan :",lieux.situ.isnull().sum())
print(lieux.shape)

taille 0 : (39296, 18)
taille nan : 943
(765576, 18)


In [87]:
# on remplace par le mode :
lieux.situ.astype(str).describe()
#top est 1 ( sur chaussé)

count     765576
unique         7
top          1.0
freq      658335
Name: situ, dtype: object

In [88]:

lieux.loc[:,['situ']]=lieux.loc[:,['situ']].replace(0,1)
lieux.loc[:,['situ']]=lieux.loc[:,['situ']].replace(np.nan,1)

###### point école :

In [89]:
# on sait pas pour cette variables quelles sont les veleurs prises
lieux.env1.unique()
# à voir ...


array([  0.,   3.,  99.,  nan])

##### variables numériques :

###### voie :

In [90]:
lieux.voie.unique()

array([41.0, 916.0, 110.0, ..., 'D2204', '885', 2913.0], dtype=object)

In [91]:
# save table to csv file
lieux.to_csv('lieux_2005-2016.csv',index=False)

---
#### la table usagers 
---

In [92]:
#shape :
usagers.shape

(1710043, 12)

In [93]:
print(usagers.dtypes)

Num_Acc      int64
place      float64
catu         int64
grav         int64
sexe         int64
trajet     float64
secu       float64
locp       float64
actp       float64
etatp      float64
an_nais    float64
num_veh     object
dtype: object


###### place : 

In [94]:
usagers.place.unique()

array([  1.,   2.,   4.,   5.,   3.,   0.,   8.,   9.,   7.,   6.,  nan])

In [95]:
print("taille 0 :" , usagers[usagers.place==0].shape)
print("taille nan :" , usagers.place.isnull().sum())


taille 0 : (60766, 12)
taille nan : 86734


In [96]:
#on les remplace par ? 
# ici les valeurs manquantes vont dependre de la nature du vehicule 
# donc pour chaque type de vehicule on va remplacer les valeurs manquates par le mode.
# donc on verra cette variable apès avoir fait la jointure entre les tables 
#sinon on peut simplifier et remplacer par le mode:
usagers.place.astype(str).describe()

count     1710043
unique         11
top           1.0
freq      1275547
Name: place, dtype: object

In [97]:
usagers.loc[:,['place']]=usagers.loc[:,['place']].replace(0,1)
usagers.loc[:,['place']]=usagers.loc[:,['place']].replace(np.nan,1)

###### Catégorie d'usager :

In [98]:
usagers.catu.unique()
#c'est bon

array([1, 2, 3, 4], dtype=int64)

###### Gravité de l'accident​ :

In [99]:
usagers.grav.unique()
#c'est bon

array([4, 3, 1, 2], dtype=int64)

###### Sexe de l'usager :


In [100]:
usagers.sexe.unique()
#c'est bon

array([1, 2], dtype=int64)

###### Annee de naissance de l'usager'

In [101]:
usagers.an_nais.dtype
# pas de soucis ici sauf si il y une erreur dans la saisie  d'une data :
print('min des dates :',usagers.an_nais.min())
print('max des dates :',usagers.an_nais.max())

min des dates : 1896.0
max des dates : 2016.0


Si les accidents ont lieu entre 2005-2016, alors un passer né en 1896 il a eu à la date d'accident en minimum un âge de 109 
et au max 120 ans. Ces deux date restent acceptable ~~

###### Motif du déplacement au moment de l’accident :

In [102]:
usagers.trajet.unique()

array([  1.,   3.,   0.,   5.,   9.,   4.,   2.,  nan])

In [103]:
print("taille 0 :" , usagers[usagers.trajet==0].shape)
print("taille nan :" , usagers.trajet.isnull().sum())

taille 0 : (506839, 12)
taille nan : 272


In [104]:
# on remplace par 9 :
usagers.loc[:,['trajet']]=usagers.loc[:,['trajet']].replace(0,9)
usagers.loc[:,['trajet']]=usagers.loc[:,['trajet']].replace(np.nan,9)

###### secu :

In [105]:
usagers.secu.unique()

array([ 11.,  31.,  21.,   0.,  12.,  13.,  90.,  10.,  22.,  91.,  33.,
        41.,  23.,  20.,  92.,  32.,  43.,  93.,  42.,   1.,  nan,  30.,
        40.,   3.,   2.])

In [125]:
secu_values=list(usagers.secu.astype(str).values)


In [126]:
# on traite d'abord le len puis les nan :
secu_values_set=[secu[:-2] for secu in secu_values if secu!='nan']
# d'après le document chaque valeur est censée d'avoir deux chiffres:

In [127]:
print(set(secu_values_set))
# il y le 0 et parfois on a qu'un seul chiffre ( ex 2, 1, 3)

{'90', '32', '13', '40', '12', '41', '33', '93', '2', '21', '23', '43', '42', '1', '31', '20', '10', '22', '30', '11', '3', '92', '91'}


In [123]:
usagers.secu.astype(str).describe()
#top 11 : ceinture exist et elle est utilisée

count     1710043
unique         25
top          11.0
freq       952293
Name: secu, dtype: object

In [124]:
# remplacer nan et 0 par 11 :
usagers.loc[:,['secu']]=usagers.loc[:,['secu']].replace(0,11)
usagers.loc[:,['secu']]=usagers.loc[:,['secu']].replace(np.nan,11)

In [132]:
# pour le reste : soit on simplifie et on remplace par 11
# soit on compléte le chiffre par un deuxième qui soit le mode approprié au chiffre existant :
# deuxième méthode est plus logique :
from statistics import mode
secu_1=[secu for secu in secu_values if '1' in secu]
mode(secu_1)
# donc on remplace par 11


'11.0'

In [133]:
secu_2=[secu for secu in secu_values if '2' in secu]
mode(secu_2)
# on remplace par '21'

'21.0'

In [134]:
secu_3=[secu for secu in secu_values if '3' in secu]
mode(secu_3)
# on remplace par '13'

'13.0'

In [135]:
usagers.loc[:,['secu']]=usagers.loc[:,['secu']].replace(1,11)
usagers.loc[:,['secu']]=usagers.loc[:,['secu']].replace(2,21)
usagers.loc[:,['secu']]=usagers.loc[:,['secu']].replace(3,13)


###### Localisation du piéton :

In [138]:
usagers.locp.unique()

array([  0.,   2.,   4.,   1.,   5.,   6.,   3.,   8.,   7.,  nan])

In [140]:
print("taille 0 :" , usagers[usagers.locp==0].shape)
print("taille nan :" , usagers.locp.isnull().sum())
print(usagers.shape)

taille 0 : (1579313, 12)
taille nan : 1071
(1710043, 12)


In [141]:
# ici on va laisser 0 comme une autre valeur, il se peut qu'elle aie pour signification : 'il n'y a pas de piéton'
# on remplace nan par 0 également 
usagers.loc[:,['locp']]=usagers.loc[:,['locp']].replace(np.nan,0)

###### Action du piéton :

In [142]:
usagers.actp.unique()

array([  0.,   3.,   1.,   5.,   2.,   9.,   6.,   4.,  nan])

In [143]:
# nan ==> 0
usagers.loc[:,['locp']]=usagers.loc[:,['locp']].replace(np.nan,0)

###### etatp

In [144]:
usagers.etatp.unique()

array([  0.,   2.,   1.,   3.,  nan])

In [145]:
# on laisse 0 comme une autre valeur qui va indiquer qu'il n'y avait pas de piéton
# nan ==> 0
usagers.loc[:,['etatp']]=usagers.loc[:,['etatp']].replace(np.nan,0)

Save to csv file :

In [146]:
usagers.to_csv('usager_2005-2016.csv', index=False)

---
la table des vehicules
---

In [148]:
vehicules.shape

(1307042, 9)

In [149]:
vehicules.dtypes


Num_Acc      int64
senc       float64
catv         int64
occutc       int64
obs        float64
obsm       float64
choc       float64
manv       float64
num_veh     object
dtype: object

###### Sens de circulation

In [147]:
vehicules.senc.unique()

array([  0.,   1.,   2.,  nan])

In [150]:
print("taille 0 :" , vehicules[vehicules.senc==0].shape)
print("taille nan :" , vehicules.senc.isnull().sum())

taille 0 : (1199403, 9)
taille nan : 72


In [151]:
vehicules.senc.astype(str).describe()

count     1307042
unique          4
top           0.0
freq      1199403
Name: senc, dtype: object

In [152]:
# on remplace par une valeur au choix :  sens croissant 
vehicules.loc[:,['senc']]=vehicules.loc[:,['senc']].replace(np.nan,1)
vehicules.loc[:,['senc']]=vehicules.loc[:,['senc']].replace(0,1)

###### Catégorie du véhicule :

In [154]:
print(len(vehicules.catv.unique()))
vehicules.catv.unique()
#Ok

33


array([ 7,  2,  9,  5, 18, 17,  1, 10,  4, 15, 19, 14, 13, 12, 21, 20, 99,
        3, 16,  8,  6, 11, 38, 32, 37, 30, 36, 31, 33, 39, 34, 35, 40], dtype=int64)

###### Obstacle mobile heurté :

In [155]:
vehicules.obsm.unique()

array([  2.,   0.,   1.,   9.,   4.,   5.,   6.,  nan])

In [156]:
# nan,0===>9
vehicules.loc[:,['obsm']]=vehicules.loc[:,['obsm']].replace(np.nan,9)
vehicules.loc[:,['obsm']]=vehicules.loc[:,['obsm']].replace(0,9)

###### Point de choc initial :

In [157]:
vehicules.choc.unique()

array([  1.,   8.,   7.,   0.,   3.,   2.,   6.,   4.,   5.,   9.,  nan])

In [158]:
# 0, nan ==> mode :
vehicules.choc.astype(str).describe()

count     1307042
unique         11
top           1.0
freq       485214
Name: choc, dtype: object

In [159]:
vehicules.loc[:,['choc']]=vehicules.loc[:,['choc']].replace(np.nan,1)
vehicules.loc[:,['choc']]=vehicules.loc[:,['choc']].replace(0,1)

###### Nombre d’occupants dans le transport en commun

In [160]:
vehicules.occutc.unique()
# 0 pour dire que ce n'est pas transport commun.

array([  0,   1,   2,  26,  23,  36,  31,   6,  51,  27,  52,  25,  10,
         8,  14,  20,  11,  40,  33,  12,   3,  13,  45,  39,   5,  15,
        32, 100,  30,  29,  35,  41,   4,  21,  16,  18,  37,  60,  62,
        46,  50,  28,  44,   7,  57,  42,   9,  17,  58,  24,  43,  48,
        22, 470,  69,  61, 150,  47, 120,  66,  49,  38,  54,  59, 200,
        34,  53,  55,  96,  65, 210,  63,  93,  80,  85, 213,  64, 193,
       111,  70, 700, 300,  73,  74,  19,  72, 110,  97,  56,  75, 220,
        81,  95,  84,  83,  67,  76,  89,  90, 480, 140,  99, 240,  71,
       132, 125, 153, 184, 170, 130, 490, 250], dtype=int64)

###### manv

In [162]:
# manv de type int
vehicules.manv.isnull().sum()

227

In [163]:
vehicules.manv.astype(str).describe()

count     1307042
unique         26
top           1.0
freq       575354
Name: manv, dtype: object

In [164]:
vehicules.loc[:,['manv']]=vehicules.loc[:,['manv']].replace(np.nan,1)

save to csv file 

In [165]:
vehicules.to_csv('vehicules_2005-2016.csv', index=False)

---
Fin nettoyage 
---

### La jointure des tables :

In [166]:
vehicules.head(5)

Num_Acc  senc  catv  occutc  obs  obsm  choc  manv num_veh
0  200500000001   1.0     7       0  0.0   2.0   1.0   1.0     A01
1  200500000001   1.0     7       0  0.0   2.0   8.0  10.0     B02
2  200500000002   1.0     7       0  0.0   2.0   7.0  16.0     A01
3  200500000002   1.0     2       0  0.0   2.0   1.0   1.0     B02
4  200500000003   1.0     2       0  0.0   2.0   1.0   1.0     A01

In [169]:
usagers.head(6)

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  200500000001    1.0     1     4     1     1.0  11.0   0.0   0.0    0.0   
1  200500000001    1.0     1     3     2     3.0  11.0   0.0   0.0    0.0   
2  200500000001    2.0     2     1     1     9.0  11.0   0.0   0.0    0.0   
3  200500000001    4.0     2     1     1     9.0  31.0   0.0   0.0    0.0   
4  200500000001    5.0     2     1     1     9.0  11.0   0.0   0.0    0.0   
5  200500000001    3.0     2     1     2     9.0  11.0   0.0   0.0    0.0   

   an_nais num_veh  
0   1976.0     A01  
1   1968.0     B02  
2   1964.0     B02  
3   2004.0     B02  
4   1998.0     B02  
5   1991.0     B02

In [172]:
print(lieux.shape)
print(caracteristiques.shape) 

(765576, 18)
(765576, 16)


on commence par ces deux table. PK/FK est Num_Acc

In [183]:
# je choisi que 90 000 ligne pour le test ( et d'ailler ca passe pas la taille entière sur mon pc)
carac_lieux_marged=pd.merge(caracteristiques.iloc[:90000,:],lieux.iloc[:90000,:],on='Num_Acc')

In [198]:
carac_lieux_marged.head()

Num_Acc  an  mois  jour  hrmn  lum  agg  int  atm  col ...     pr1  \
0  200500000001   5     1    12  1900    3    2    1  1.0  3.0 ...   430.0   
1  200500000002   5     1    21  1600    1    2    1  1.0  1.0 ...     0.0   
2  200500000003   5     1    21  1845    3    1    1  2.0  1.0 ...     0.0   
3  200500000004   5     1     4  1615    1    1    1  1.0  5.0 ...     0.0   
4  200500000005   5     1    10  1945    3    1    1  3.0  6.0 ...   630.0   

  vosp prof  plan  lartpc  larrout  surf infra  situ env1  
0  NaN  1.0   1.0     0.0     63.0   1.0   0.0   1.0  0.0  
1  1.0  1.0   1.0     0.0    100.0   1.0   0.0   5.0  0.0  
2  1.0  1.0   1.0     0.0      0.0   2.0   0.0   5.0  0.0  
3  NaN  1.0   1.0     0.0      0.0   1.0   0.0   1.0  0.0  
4  NaN  1.0   3.0     0.0     59.0   2.0   0.0   3.0  0.0  

[5 rows x 33 columns]

on merge avec vehicules : 

In [185]:
vehicules.head()

Num_Acc  senc  catv  occutc  obs  obsm  choc  manv num_veh
0  200500000001   1.0     7       0  0.0   2.0   1.0   1.0     A01
1  200500000001   1.0     7       0  0.0   2.0   8.0  10.0     B02
2  200500000002   1.0     7       0  0.0   2.0   7.0  16.0     A01
3  200500000002   1.0     2       0  0.0   2.0   1.0   1.0     B02
4  200500000003   1.0     2       0  0.0   2.0   1.0   1.0     A01

In [186]:
carac_lieux_veh_merged=pd.merge(carac_lieux_marged, vehicules,on='Num_Acc' )

In [188]:
carac_lieux_veh_merged.head()

Num_Acc  an  mois  jour  hrmn  lum  agg  int  atm  col   ...    situ  \
0  200500000001   5     1    12  1900    3    2    1  1.0  3.0   ...     1.0   
1  200500000001   5     1    12  1900    3    2    1  1.0  3.0   ...     1.0   
2  200500000002   5     1    21  1600    1    2    1  1.0  1.0   ...     5.0   
3  200500000002   5     1    21  1600    1    2    1  1.0  1.0   ...     5.0   
4  200500000003   5     1    21  1845    3    1    1  2.0  1.0   ...     5.0   

  env1 senc  catv  occutc  obs  obsm choc  manv num_veh  
0  0.0  1.0     7       0  0.0   2.0  1.0   1.0     A01  
1  0.0  1.0     7       0  0.0   2.0  8.0  10.0     B02  
2  0.0  1.0     7       0  0.0   2.0  7.0  16.0     A01  
3  0.0  1.0     2       0  0.0   2.0  1.0   1.0     B02  
4  0.0  1.0     2       0  0.0   2.0  1.0   1.0     A01  

[5 rows x 41 columns]

In [192]:
usagers.shape

(1710043, 12)

In [197]:
usagers.head()

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  200500000001    1.0     1     4     1     1.0  11.0   0.0   0.0    0.0   
1  200500000001    1.0     1     3     2     3.0  11.0   0.0   0.0    0.0   
2  200500000001    2.0     2     1     1     9.0  11.0   0.0   0.0    0.0   
3  200500000001    4.0     2     1     1     9.0  31.0   0.0   0.0    0.0   
4  200500000001    5.0     2     1     1     9.0  11.0   0.0   0.0    0.0   

   an_nais num_veh  
0   1976.0     A01  
1   1968.0     B02  
2   1964.0     B02  
3   2004.0     B02  
4   1998.0     B02

on merge finalement avec usagers :

In [200]:
#  later 
final_data=pd.merge(carac_lieux_veh_merged,usagers,on=['num_veh','num_veh'])

MemoryError: 